src/DevBetterWeb.Core/Entities/Invitation.cs

```diff
@@ -9,12 +10,27 @@ public class Invitation : BaseEntity, IAggregateRoot
     public string InviteCode { get; private set; }
     public string PaymentHandlerSubscriptionId { get; private set; }
     public bool Active { get; private set; } = true;
+    public DateTime DateCreated { get; private set; }
+    public DateTime DateOfUserPing { get; private set; } = DateTime.MinValue;
+    public DateTime DateOfLastAdminPing { get; private set; } = DateTime.MinValue;
 
     public Invitation(string email, string inviteCode, string paymentHandlerSubscriptionId)
     {
       Email = email;
       InviteCode = inviteCode;
       PaymentHandlerSubscriptionId = paymentHandlerSubscriptionId;
+      DateCreated = DateTime.Today;
+    }
+
+    // for testing
```

Comment:
Why do you need a separate constructor for testing? That's usually a code smell. Can you not configure the state of the type using its regular constructor plus properties?

Reply:
I needed to test logic based on when the various events (creation, user ping, admin ping) happened but those shouldn't really be able to be publically set. Especially creation, which (except in testing) should just happen in the constructor. Definitely a code smell but I'm not sure of a better way to do it as I was trying to avoid making these things have public setters

Reply:
Something else I missed - don't set things to DateTime.MinValue as defaults or to mean "never happened". Instead, use a nullable DateTime (`DateTime?`) and default them to null. If that makes it complain about nullable reference types you can override the warning.

If creation always defaults to Today, that should be easily unit tested. There would only be rare edge cases where you ran the tests near midnight when a failure would occur. So no need to have a separate way to set that, typically. Instead you might have a way of abstracting time in your service, so that even if the invite is created Today, the service might be executing on Today.AddDays(5) which you could specify via an IDateTime mock that is injected into the service, for instance.

The other dates must have some way of being set so probably you could call whatever method sets them to put them into a valid state. Having a separate test constructor works but is usually a last resort.